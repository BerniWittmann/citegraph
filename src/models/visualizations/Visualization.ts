import Transformer from '@/transformers/Transformer'

export interface TimePeriod {
  from: number
  to: number
}

export interface VisualizationBaseFields {
  id?: string;
  progress?: number;
  name: string;
  timePeriods?: Array<TimePeriod>
}

export interface VisualizationBaseFieldsDTO {
  id?: string;
  progress?: number;
  name: string;
  key: string;
  // eslint-disable-next-line camelcase
  time_periods?: Array<TimePeriod>
}

export default abstract class Visualization implements VisualizationBaseFields {
  id?: string;
  progress?: number;
  name: string;
  timePeriods: Array<TimePeriod>
  abstract parameters: Object;
  abstract data?: Object;
  abstract readonly key: string;
  static readonly key: string;
  static readonly title: string;
  static readonly imageUrl: string;
  static readonly description: string;
  static readonly longDescription: string;
  static readonly transformer: typeof Transformer;

  protected constructor ({ id, progress, name, timePeriods }: VisualizationBaseFields) {
    this.id = id
    this.progress = progress
    this.name = name
    this.timePeriods = timePeriods || []
  }
}
