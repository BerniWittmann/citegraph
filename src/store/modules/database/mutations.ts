import { DatabaseState } from './types'
import { MutationTree } from 'vuex'

export const mutations: MutationTree<DatabaseState> = {}
