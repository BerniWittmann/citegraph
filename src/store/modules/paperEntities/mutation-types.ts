export const SET_ENTITIES = 'SET_ENTITIES'
export const SET_ENTITY_TYPE = 'SET_ENTITY_TYPE'
export const SET_ENTITY_COUNT = 'SET_ENTITY_COUNT'
