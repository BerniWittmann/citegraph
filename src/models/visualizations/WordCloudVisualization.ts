import Visualization, { VisualizationBaseFields } from './Visualization'

export interface WordCloudVisualizationParameters {
  pruneLeastCitedPercentage: number
}

export interface WordCloudDataPoint {
  word: string
  weight: number
}
export interface WordCloudData {
  data: Array<WordCloudDataPoint>
}

export interface WordCloudVisualizationFields {
  data?: WordCloudData
}

export default class WordCloudVisualization extends Visualization implements WordCloudVisualizationFields {
  static key: string = 'wordcloud'
  key: string = 'wordcloud'
  static title: string = 'visualizations.information.word_cloud.title'
  static description: string = 'visualizations.information.word_cloud.description'
  static longDescription: string = 'visualizations.information.word_cloud.long_description'
  static imageUrl: string = 'visualization_images/word_cloud.jpg'
  parameters: WordCloudVisualizationParameters
  data?: WordCloudData

  constructor ({ id, name, progress, data }: VisualizationBaseFields & WordCloudVisualizationFields) {
    super({ id, name, progress })
    this.parameters = {
      pruneLeastCitedPercentage: 5
    }
    this.data = data
  }
}
