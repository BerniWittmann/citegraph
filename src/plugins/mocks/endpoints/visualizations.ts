import { AxiosRequestConfig } from 'axios'
import MockAdapter from 'axios-mock-adapter'
import db from '../db'
import Project from '@/models/project'
import Visualization from '@/models/visualizations/Visualization'

const projects = db.getCollection('projects')
const visualizations = db.getCollection('visualizations')

function getProjectId (config: AxiosRequestConfig): number {
  const urlParts: Array<string> = config.url!.split('/')
  return parseInt(urlParts[4])
}

function getVisualizationId (config: AxiosRequestConfig): string {
  const urlParts: Array<string> = config.url!.split('/')
  return urlParts[6]
}

export function setupMocks (mock: MockAdapter) {
  mock.onGet(/\/projects\/\d+\/visualizations$/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    return [200, visualizations.find()]
  })

  mock.onGet(/\/projects\/\d+\/visualizations\/\d+$/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    const visualization: Resultset<Visualization> = visualizations.findOne({ 'id': getVisualizationId(config) })
    if (!visualization) {
      return [404]
    }
    return [200, visualization]
  })
}
