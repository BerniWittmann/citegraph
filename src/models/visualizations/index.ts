import WordCloudVisualization, { WordCloudVisualizationFields } from './WordCloudVisualization'
import { VisualizationBaseFields } from '@/models/visualizations/Visualization'
import BarChartVisualization, { BarChartVisualizationFields } from '@/models/visualizations/BarChartVisualization'

export const visualizations = [WordCloudVisualization, BarChartVisualization]

const visualizationsKeyMap: Record<string, any> = {}
visualizations.forEach((visualization) => {
  visualizationsKeyMap[visualization.key] = visualization
})

export {
  visualizationsKeyMap
}

export type VisalizationFields = VisualizationBaseFields & (
  WordCloudVisualizationFields &
  BarChartVisualizationFields
)
