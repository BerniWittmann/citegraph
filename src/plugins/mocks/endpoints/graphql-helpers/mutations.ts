import { PaperEntityFields } from '@/models/paperEntities'
import { QueryConfiguration } from '@/plugins/mocks/endpoints/graphql-helpers/queries'
import db from '../../db'

type MutationArguments = {
  id?: string
}

function parseQuery (query: string): MutationArguments {
  const idString = query.match(/id: [a-zA-Z0-9]+/)
  const id = idString ? idString[0].toString().substring(4) : undefined
  return {
    id
  }
}

function parseData (query: string): Partial<PaperEntityFields> {
  // extract update data
  const result: Record<string, any> = {}
  const items = query.match(/\w+: ".*"/g) || []
  items.forEach((item: string) => {
    const key = item.split(': ')[0]
    const val = item.split(': ')[1]
    result[key] = val.substring(1, val.length - 1).replace(/\/"/g, '')
  })
  return result
}

function updateOne (queryParams: MutationArguments, data: Partial<PaperEntityFields>, configuration: QueryConfiguration) {
  let entity = configuration.collection.findOne({ id: queryParams.id })
  if (!entity) {
    throw new Error('Entity not found')
  }
  for (let key in data) {
    // @ts-ignore
    entity[key] = data[key]
  }
  return configuration.collection.update(entity)
}

export function createMutationResponse (query: string, configuration: QueryConfiguration) {
  const queryParams = parseQuery(query)
  if (!queryParams.id) {
    return [400, 'No id given']
  }
  try {
    const response = updateOne(queryParams, parseData(query), configuration)
    db.saveDatabase()
    return [200, {
      data: {
        [configuration.queryName]: {
          ...response
        }
      }
    }]
  } catch (err) {
    return [404, err]
  }
}
