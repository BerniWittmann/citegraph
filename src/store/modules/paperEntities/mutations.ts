import { PaperEntitiesState } from './types'
import { MutationTree } from 'vuex'
import {
  SET_ACTIVE_ENTITY,
  SET_ENTITIES,
  SET_ENTITY_COUNT,
  SET_ENTITY_TYPE
} from '@/store/modules/paperEntities/mutation-types'
import { PaperEntity } from '@/models/paperEntities'

export const mutations: MutationTree<PaperEntitiesState> = {
  [SET_ENTITIES] (state: PaperEntitiesState, payload: Array<PaperEntity>) {
    state.entities = payload
  },
  [SET_ENTITY_TYPE] (state: PaperEntitiesState, payload: string | undefined) {
    state.entityType = payload
  },
  [SET_ENTITY_COUNT] (state: PaperEntitiesState, payload: number) {
    state.entityCount = payload
  },
  [SET_ACTIVE_ENTITY] (state: PaperEntitiesState, payload: PaperEntity | undefined) {
    state.activeEntity = payload
  }
}
