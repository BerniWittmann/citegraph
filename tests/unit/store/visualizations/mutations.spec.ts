import { mutations } from '@/store/modules/visualizations/mutations'
import * as mutationTypes from '@/store/modules/visualizations/mutation-types'
import Visualization from '@/models/visualizations/Visualization'
import { VisualizationsState } from '@/store/modules/visualizations/types'
import WordCloudVisualization from '@/models/visualizations/WordCloudVisualization'

describe('store/modules/visualizations/mutations', () => {
  it('provides the mutations', () => {
    expect(mutations).toMatchSnapshot()
  })

  describe('SET_VISUALIZATIONS', () => {
    it('saves the visualizations to the store', () => {
      const state: VisualizationsState = {
        visualizations: []
      }
      const visualizations: Array<Visualization> = [
        new WordCloudVisualization({ id: '1', name: 'First Project' }),
        new WordCloudVisualization({ id: '2', name: 'Second Project' })
      ]

      mutations[mutationTypes.SET_VISUALIZATIONS](state, visualizations)

      expect(state.visualizations).toEqual(visualizations)
    })

    it('overrides existing projects', () => {
      const state: VisualizationsState = {
        visualizations: [
          new WordCloudVisualization({ id: '11', name: 'Other Project' }),
          new WordCloudVisualization({ id: '21', name: 'Old Project' })
        ]
      }
      const visualizations: Array<Visualization> = [
        new WordCloudVisualization({ id: '1', name: 'First Project' }),
        new WordCloudVisualization({ id: '2', name: 'Second Project' })
      ]

      mutations[mutationTypes.SET_VISUALIZATIONS](state, visualizations)

      expect(state.visualizations).toEqual(visualizations)
    })
  })

  describe('SET_CURRENT_VISUALIZATION', () => {
    it('saves the visualization to the store', () => {
      const state: VisualizationsState = {
        visualizations: [],
        currentVisualization: undefined
      }
      const visualization: Visualization = new WordCloudVisualization({ id: '1', name: 'First Project' })

      mutations[mutationTypes.SET_CURRENT_VISUALIZATION](state, visualization)

      expect(state.currentVisualization).toEqual(visualization)
    })

    it('overrides existing visualization', () => {
      const state: VisualizationsState = {
        visualizations: [],
        currentVisualization: new WordCloudVisualization({ id: '42', name: 'Other Project' })
      }
      const visualization: Visualization = new WordCloudVisualization({ id: '1', name: 'First Project' })

      mutations[mutationTypes.SET_CURRENT_VISUALIZATION](state, visualization)

      expect(state.currentVisualization).toEqual(visualization)
    })

    it('can unset a visualization', () => {
      const state: VisualizationsState = {
        visualizations: [],
        currentVisualization: new WordCloudVisualization({ id: '42', name: 'Other Project' })
      }

      const visualization = undefined
      mutations[mutationTypes.SET_CURRENT_VISUALIZATION](state, visualization)

      expect(state.currentVisualization).toEqual(visualization)
    })
  })
})
