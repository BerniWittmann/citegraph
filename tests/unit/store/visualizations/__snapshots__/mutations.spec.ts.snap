// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/modules/visualizations/mutations provides the mutations 1`] = `
Object {
  "SET_CURRENT_VISUALIZATION": [Function],
  "SET_VISUALIZATIONS": [Function],
}
`;
